contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,129 @@
+
+Since `retrivePrice()` is a non-blocking function and it's only handled after the `console.log()` has executed, there must be a way to ensure that these functions are executed in the sequence that's desired.
+
+This is where Promises come in handy.
+
+
+## Introduction
+Promises in Javscript behave the same way as Promises do in real life. It is a proxy for a value that is not known at the time of Promise creation. It has 3 states: pending, fulfilled and rejected. Promises provide the power to control the execution flow of your program by allowing the user to wait for a promise to be resolved before carrying on with the rest of the program execution.
+
+## How to use Promises
+
+The Promise API is only supported natively from ES6 onwards, which is the version that modern browsers support. The code below will thus work in any modern browser and also in Node.js (from version 6). 
+
+
+Promises are created in Javascript by using the `new` constructor as follows:
+
+```javascript
+
+var myFirstPromise = new Promise(function(resolve, reject) {
+    // insert stuff to do here
+    // The promise is currently 'pending'
+})
+
+```
+
+The Promise constructor takes function that has two parameters: resolve and reject. These two parameters are functions which will change the state of the promise. They can be called with or without a parameter as follows:
+
+```javascript
+
+resolve(someValue);
+reject(someValue);
+resolve();
+reject();
+
+```
+
+## Handling results from Promises
+
+As promises have 2 completed states, there are 2 functions available to deal with these 2 scenarios. 
+
+### .then
+
+`.then` is used to deal with a Promise after it has been resolved. It takes in a function with 1 parameter, which is the result of the Promise that has just been resolved.
+
+### .catch
+
+`.catch` is used to deal with a Promise after it has been rejected. Like `.then`, it also takes in a function with 1 parameter.
+
+Whenever you want to act on a result from a Promise, it is good practice to use the above 2 functions to ensure certainty and code readability.
+
+An example would be as follows:
+
+```javascript
+
+var isSuccessful = false;
+
+var myFirstPromise = new Promise(function(resolve, reject) {
+    if (isSuccessful) {
+        resolve('Success');
+    }
+    else
+        reject('Failure');
+}).then(function(result) {
+    console.log(result);    // prints 'Success'
+}).catch(function(result) {
+    console.log(result);    // prints 'Failure'
+});
+
+```
+
+## Benefits
+
+- Avoid callback hell
+
+![Callback Hell](callbackhell.png)
+(Credit: [Brett McLain](http://blog.mclain.ca/))
+
+In Javascript, callback functions are usually used to perform some action on a value returned from another function. However, this may result in deeply nested code as shown above, especially if there are many asynchronous method calls. When using promises, the nesting is kept at a minimum and results in cleaner and readable code as shown below. 
```

Comment:
This is where you explain what a callback is, but I remember seeing the term callback near the beginning of this chapter. :-p